ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixFactory.java

```diff
@@ -69,10 +73,65 @@ public HelixManager getZkHelixManagerAndConnect(String clusterName, String insta
+   * @param metrics the {@link DataNodeConfigSourceMetrics} to use.
+   * @return either a new instance of {@link DataNodeConfigSource} with the supplied configuration, or one from the pool
+   *         if there is already one created for this address.
+   */
+  public DataNodeConfigSource getDataNodeConfigSource(ClusterMapConfig clusterMapConfig, String zkAddr,
+      DataNodeConfigSourceMetrics metrics) {
+    return dataNodeConfigSources.computeIfAbsent(zkAddr, k -> buildSource(clusterMapConfig, zkAddr, metrics));
+  }
+
+  /**
+   * @param clusterMapConfig the {@link ClusterMapConfig} to use.
+   * @param zkAddr the ZooKeeper address to connect to. If a {@link HelixManager} is required and one is already in the
+   *               pool with this address, it will be reused.
+   * @param metrics the {@link DataNodeConfigSourceMetrics} to use.
+   * @return a new instance of {@link DataNodeConfigSource} with the supplied configuration.
+   */
+  private DataNodeConfigSource buildSource(ClusterMapConfig clusterMapConfig, String zkAddr,
+      DataNodeConfigSourceMetrics metrics) {
+    try {
+      InstanceConfigToDataNodeConfigAdapter instanceConfigSource = null;
+      if (clusterMapConfig.clusterMapDataNodeConfigSourceType.isInstanceConfigAware()) {
+        instanceConfigSource = new InstanceConfigToDataNodeConfigAdapter(
+            getZkHelixManagerAndConnect(clusterMapConfig.clusterMapClusterName,
+                ClusterMapUtils.getInstanceName(clusterMapConfig.clusterMapHostName, clusterMapConfig.clusterMapPort),
+                InstanceType.SPECTATOR, zkAddr), clusterMapConfig);
+      }
+      PropertyStoreToDataNodeConfigAdapter propertyStoreSource = null;
+      if (clusterMapConfig.clusterMapDataNodeConfigSourceType.isPropertyStoreAware()) {
+        propertyStoreSource = new PropertyStoreToDataNodeConfigAdapter(zkAddr, clusterMapConfig);
+      }
+
+      switch (clusterMapConfig.clusterMapDataNodeConfigSourceType) {
+        case INSTANCE_CONFIG:
+          return instanceConfigSource;
+        case PROPERTY_STORE:
+          return propertyStoreSource;
+        case COMPOSITE_INSTANCE_CONFIG_PRIMARY:
+          return new CompositeDataNodeConfigSource(instanceConfigSource, propertyStoreSource, SystemTime.getInstance(),
+              metrics);
+        case COMPOSITE_PROPERTY_STORE_PRIMARY:
+          return new CompositeDataNodeConfigSource(propertyStoreSource, instanceConfigSource, SystemTime.getInstance(),
+              metrics);
+        default:
+          throw new IllegalArgumentException("Unknown type: " + clusterMapConfig.clusterMapDataNodeConfigSourceType);
+      }
+    } catch (Exception e) {
+      throw new RuntimeException("Exception while instantiating DataNodeConfigSource", e);
+    }
+  }
+
   /**
    * Hashable key used to cache instances of {@link HelixManager} that match desired parameters.
+   * Exposed for use in testing.
    */
-  private static class ManagerKey {
+  static class ManagerKey {
```

Comment:
nit: in the `equals` method, suggest comparing `zkAddr` and `instanceName` first.

Reply:
Compared zkAddr and instanceType first since those are most likely to be different in a single process.